src/vm/types.rs

```diff
@@ -24,45 +26,51 @@ pub enum AtomTypeIdentifier {
     TupleType(TupleTypeSignature)
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 struct ListTypeData {
     max_len: u32,
     dimension: u8
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct TypeSignature {
     atomic_type: AtomTypeIdentifier,
     list_dimensions: Option<ListTypeData>,
     // NOTE: for the purposes of type-checks and cost computations, list size = dimension * max_length!
     //       high dimensional lists are _expensive_ --- use lists of tuples!
 }
 
-#[derive(Debug, Clone, Eq)]
+#[derive(Debug, Clone, Eq, Serialize, Deserialize)]
 pub struct TupleData {
     type_signature: TupleTypeSignature,
     data_map: BTreeMap<String, Value>
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub struct BuffData {
-    data: Vec<u8>,
+    pub data: Vec<u8>,
 }
 
-#[derive(Debug, Clone, Eq)]
+#[derive(Debug, Clone, Eq, Serialize, Deserialize)]
 pub struct ListData {
     pub data: Vec<Value>,
     type_signature: TypeSignature
 }
 
-#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum PrincipalData {
+    StandardPrincipal(u8, [u8; 20]),  // a standard principal is a version byte + hash160 (20 bytes)
+    ContractPrincipal(String)
```

Comment:
Won't this be a (possibly versioned) hash160 as well?

Reply:
If contract names are user chosen, this wouldn't need to be a hash160. Unless we wanted it to be so for the wire format.

Reply:
I think the decision we reached was that contract names would be composed of the Principal that created them, and a human-given name supplied by the principal.  So a fully-qualified contract name might be something like `SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR.naming_system`.  Would this be represented as a `String`, or as a tuple like `(StandardPrincipal(u8, [u8; 20]), String)`?

Reply:
ah -- probably closer to the second, though I think we'd want to make the "StandardPrincipal" a class like `VersionedHash160`, probably defined in the address module, so we'd have an enum like:

```
StandardPrincipal(VersionedHash160),
ContractPrincipal(VersionedHash160, String)
```

Since you can't use an enum variant within another variant of the same enum.